File: 3 kyu/Prime Streaming (PG-13)/solution.java
import java.util.stream.IntStream;

public class Primes {
  public static IntStream stream() {
    return IntStream.rangeClosed(2, Integer.MAX_VALUE).filter(Primes::isPrime);
  }
  
  public static boolean isPrime(int number) {
    for (int i = 2; i < (int) Math.sqrt(number) + 1; i++)
      if (number % i == 0) return false;
  
    return true;
  }
}
___________________________________________________________
import java.util.stream.IntStream;
import java.util.function.IntSupplier;
public class Primes {
  
  public static IntStream stream() {
    return IntStream.generate(
      new IntSupplier(){
      int lastSeen = 1;
      public int getAsInt(){
      int i = lastSeen;
      while(!isPrime(++i)){}
      lastSeen = i;
      return i;
    }});
  }
  public static boolean isPrime(int i){
    if(i==1) return false;
    if(i==2 || i==3 || i==5 || i==7 || i==11 || i==13 || i==17 || i==19) return true; // edge cases
    if(i%2==0 || i%3==0) return false; // edge cases
    int endp = (int)Math.ceil(Math.sqrt(i)/6.0)+2;
    for(int m = 1; m <= endp; m++){
      if(i%(6*m+1)==0 || i%(6*m-1)==0) return false;
    }
    return true;
  }
  
}
___________________________________________________________
import java.util.stream.IntStream;

public class Primes {
  private static final int N = 20000000;
  private static final boolean[] SIEVE = new boolean[N + 1];
  static {
    for (int i = 2; i * i <= N; i++) {
      if (!SIEVE[i]) {
        for (int j = i; i * j <= N; j++) {
          SIEVE[i * j] = true;
        }
      }
    }
  }

  public static IntStream stream() {
    return IntStream.range(2, N).filter(x -> !SIEVE[x]);
  }
}
___________________________________________________________
import java.util.stream.IntStream;

public class Primes {
  public static IntStream stream() {
        return IntStream.concat(
                IntStream.of(2),
                IntStream.iterate(3, n -> n + 2).filter(Primes::isPrime)
        );
    }

    private static boolean isPrime(int n) {
        return IntStream.iterate(3, x -> x <= (int) Math.sqrt(n), x -> x + 2)
                .allMatch(i -> n % i != 0);
    }
}

File: 3 kyu/Prime Streaming (PG-13)/solution.js
class Primes {
  static * stream() {
  yield 2 
  var n = 3
   while (true) {
     if (isPrime(n)) {
       yield n
     }
     n += 2
   }
  }
}

function isPrime(n) {
  for (var a = 3; a <= ~~Math.sqrt(n); a+=2) {
    if (n%a == 0) return false;
    }
  return true;
}
___________________________________________________________
class Primes {
  static *stream() {
      yield 2; yield 3; yield 5; yield 7;
      let map = new Map();
      let gen = Primes.stream();
      gen.next() && gen.next();
      let x = 3; let i = 7;
      while (true) {
        i+=2;
        let val = map.get(i) || -1;
        if (val > -1) map.delete(i);
        else if (i < x*x){
          yield i;
          continue;
        }
        else {
          val = x*2;
          x = gen.next()['value'];
        }
        let key = i + val;
        while (map.has(key)) key+=val;
        map.set(key, val);
      }
  }
}
___________________________________________________________
limit = 15495863
const sieve = [0,0].concat(new Array(limit-2).fill(1))
const sl = sieve.length
const already_set = []
class Primes {
  static * stream() {
      for (let i=0;i<sl;i++){
        if (sieve[i]){
           yield i
           if (!already_set[i]){
             already_set[i]==1
             for (let k = i*i;k<limit;k+=i) sieve[k] = 0
            }
        }
      }
  }
}
___________________________________________________________
const isPrime = function(n) {
  if (n <= 3) {
    return n > 1;
  } else if (n % 2 === 0 || n % 3 === 0) {
    return false;
  }
  
  let i = 5;
  const sqrt = parseInt(Math.sqrt(n));
  while (i <= sqrt) {
    if (n % i === 0 || n % (i + 2) === 0)
      return false;
    i += 6;
  }
  
  return true;
}

class Primes {
  static * stream() {
    let i = 2;
    while (true) {
      if (isPrime(i))
        yield i;
      i++;
    }
  }
}
